pkg/plank/reconciler.go

```diff
@@ -672,6 +656,55 @@ func (r *reconciler) syncPendingJob(ctx context.Context, pj *prowv1.ProwJob) (*r
 	return nil, nil
 }
 
+type PodTransientFailure string
+
+const (
+	PodTransientFailureNone        PodTransientFailure = ""
+	PodTransientFailureUnknown     PodTransientFailure = "unknown"
+	PodTransientFailureEvicted     PodTransientFailure = "evicted"
+	PodTransientFailureTerminated  PodTransientFailure = "terminated"
+	PodTransientFailureUnreachable PodTransientFailure = "unreachable"
+)
+
+func getTransientFailure(pod *corev1.Pod) PodTransientFailure {
+	if pod.Status.Reason == Evicted {
+		return PodTransientFailureEvicted
+	}
+
+	// If there was a Graceful node shutdown, the Pod's status will have a
+	// reason set to "Terminated":
+	// https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown
+	if pod.Status.Reason == Terminated {
+		return PodTransientFailureTerminated
+	}
+
+	for _, condition := range pod.Status.Conditions {
+		// If the node does no longer exist and the pod gets garbage collected,
+		// this condition will be set:
+		// https://kubernetes.io/docs/concepts/workloads/pods/disruptions/#pod-disruption-conditions
+		if condition.Reason == "DeletionByPodGC" {
+			return PodTransientFailureTerminated
+		}
+
+		// On GCP, before a new spot instance is started, the old pods are garbage
+		// collected (if they have not been already by the Kubernetes PodGC):
+		// https://github.com/kubernetes/cloud-provider-gcp/blob/25e5dcc715781316bc5e39f8b17c0d5b313453f7/cmd/gcp-controller-manager/node_csr_approver.go#L1035-L1058
+		if condition.Reason == "DeletionByGCPControllerManager" {
```

Comment:
Since this is GCP related only, why not make the reasons configurable too, to allow the user to add more cases based on the infrastructure?

Reply:
I made the settings configurable: https://github.com/kubernetes-sigs/prow/pull/117/commits/648600a808d9f69de344253e1667fbd13ebad5a8